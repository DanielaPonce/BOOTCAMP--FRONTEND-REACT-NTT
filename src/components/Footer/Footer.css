/* Footer */
footer {
	background-color: var(--white);
	color: var(--text-color);
	padding: 1.5rem;
	text-align: center;
}

footer .social-icons {
	margin: 1rem 0;
}

.social-icons img {
	cursor: pointer;
}

.social-icons img:hover {
	filter: invert(50%) sepia(90%) saturate(1000%) hue-rotate(350deg)
		brightness(95%) contrast(90%);
}

footer .social-icons a:hover {
	color: var(--secondary-color);
}
