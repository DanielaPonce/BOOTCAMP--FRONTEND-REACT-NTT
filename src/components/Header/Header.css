.header-container {
	background-color: var(--primary-color);
	position: relative;
	height: 12vh;
}

.header-container header {
	position: absolute;
	top: 5vh;
	width: 90vw;
	left: 50%; 

	transform: translate(-50%, 7%); 
	box-shadow: 5px 5px 15px rgba(0, 0, 0, 22%);
	background-color: var(--white);
	border-radius: 25px;
	color: var(--text-color);
	padding: 2em 24px;

	display: flex;
	justify-content: space-between;
	align-items: center;
}

header .search-container {
	display: flex;
	justify-content: center;
	align-items: center;
	gap: 8px;
}

.logo {
	display: flex;
	align-items: center;
	gap: 0.5rem;
	text-decoration: none;
	color: var(--text-color);
	cursor: pointer;
}

.logo h1:hover {
	color: var(--hover-button-color);
}


@media only screen and (min-width: 740px) and (max-width: 991px) {
	.logo h1 {
		font-size: 24px;
	}
	
}

@media only screen and (max-width: 739px) {
	.header-container header {
		flex-direction: column;
		top: 1vh;
		padding: 16px;
		gap: 16px;
		width: 85vw;
	}

	.logo h1 {
		margin: 0;
		font-size: 24px;
	}
}
