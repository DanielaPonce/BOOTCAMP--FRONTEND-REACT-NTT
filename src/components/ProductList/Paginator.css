.paginator-container {
	display: flex;
	align-items: center;
	column-gap: 16px;
}

.paginator-container button {
	box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
	background-color: var(--white);
	border-radius: 50%;
    border: unset;
	padding: 4px 8px;
	cursor: pointer;
    transition: all 0.3s;
}

.paginator-container button:hover {
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3);
}