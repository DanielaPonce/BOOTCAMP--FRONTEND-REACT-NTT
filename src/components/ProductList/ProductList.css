.product-container {
	display: grid;
	grid-template-columns: repeat(4, 1fr);
	justify-content: center;
	gap: 3rem 1.5rem;
}

@media only screen and (min-width: 992px) and (max-width: 1439px) {
	.product-container {
		grid-template-columns: 1fr 1fr 1fr;
	}
}

@media only screen and (min-width: 740px) and (max-width: 991px) {
	.product-container {
		grid-template-columns: 1fr 1fr;
	}
}

@media only screen and (max-width: 739px) {
	.product-container {
		grid-template-columns: 1fr;
		gap: 16px;
	}
}