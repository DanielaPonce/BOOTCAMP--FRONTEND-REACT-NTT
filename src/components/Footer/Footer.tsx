import './Footer.css'

const Footer = () => {
	return (
		<footer>
			<p>Síguenos en nuestras redes sociales</p>
			<div className="social-icons">
				<img src="/src//assets/icons/fb-icon.svg" alt="facebook icon" />
				<img
					src="/src/assets/icons/instagram-icon.svg"
					alt="instagram icon"
				/>
				<img
					src="/src/assets/icons/twitter-icon.svg"
					alt="twitter icon"
				/>
			</div>
			<p>© 2024 My Market. Todos los derechos reservados.</p>
			<p>
				<a href="#">Política de privacidad</a> |
				<a href="#">Términos de uso</a> | <a href="#">Contacto</a>
			</p>
		</footer>
	);
};

export default Footer;
