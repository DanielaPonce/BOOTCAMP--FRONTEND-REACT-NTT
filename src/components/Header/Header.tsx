import { FC } from 'react';
import { Link, useLocation } from 'react-router';
import { useCategories } from '../../hooks/useCategories';
import { StaticTexts } from '../../utils/staticTexts';
import CartCounter from './CartCounter';
import CategoryDropdown from './CategoryDropdown';
import SearchBar from './SearchBar';
import { RoutesConstants } from '../../utils/routes';
import { Images } from '../../utils/images';
import './Header.css';
import { useAuth } from '../../context/AuthContext';
import Button from '../Button/Button';

type HeaderProps = {
	onSearch?: (query: string) => void;
	onFilterCategory?: (category: string) => void;
};

const Header: FC<HeaderProps> = ({ onSearch, onFilterCategory }) => {
	const { user, logout } = useAuth();
	const location = useLocation();

	const { categories } = useCategories();

	const hideSearchAndDropdown = location.pathname === '/summary';

	return (
		<div className="header-container">
			<header>
				<div className="left-side">
					<Link to={RoutesConstants.Home} className="logo">
						<img
							src={Images.Logo}
							alt="Logo"
							width={45}
							height={45}
						/>
						<h1>{StaticTexts.AppTitle}</h1>
					</Link>
					{user && <span>Bienvenido: {user}</span>}
				</div>

				<div className="search-container">
					{hideSearchAndDropdown ? (
						<></>
					) : (
						<>
							<SearchBar onInputSearch={onSearch!} />
							<CategoryDropdown
								categories={categories}
								onSelectCategory={onFilterCategory!}
							/>
						</>
					)}
					<CartCounter />
					<Button
						type="button"
						className="logout-button"
						onClick={logout}
					>
						Cerrar Sesi√≥n
					</Button>
				</div>
			</header>
		</div>
	);
};

export default Header;
