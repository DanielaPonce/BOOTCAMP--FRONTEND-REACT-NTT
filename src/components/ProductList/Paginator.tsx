import './Paginator.css';

type PaginatorProps = {
	currentPage: number;
	totalPages: number;
	onPageChange: (page: number) => void;
};

const Paginator: React.FC<PaginatorProps> = ({
	currentPage,
	totalPages,
	onPageChange
}) => {
	return (
		<div className="paginator-container">
			<button
				onClick={() => onPageChange(currentPage - 1)}
				disabled={currentPage === 1}
			>
				&lt;
			</button>
			<div id="paginator-content__number">
				Page {currentPage} of {totalPages}
			</div>
			<button
				onClick={() => onPageChange(currentPage + 1)}
				disabled={currentPage === totalPages}
			>
				&gt;
			</button>
		</div>
	);
};

export default Paginator;
