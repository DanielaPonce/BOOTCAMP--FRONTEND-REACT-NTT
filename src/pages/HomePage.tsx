import { FC, useState } from 'react';
import Footer from '../components/Footer/Footer';
import Header from '../components/Header/Header';
import Paginator from '../components/ProductList/Paginator';
import ProductList from '../components/ProductList/ProductList';
import Spinner from '../components/Spinner/Spinner';
import { CartProvider } from '../context/CartContext';
import { useProducts } from '../hooks/useProducts';
import { Constants } from '../utils/constants';

const HomePage: FC = () => {
	const [search, setSearch] = useState('');
	const [category, setCategory] = useState(
		Constants.defaultSelectValue.valueOf()
	);

	const {
		products,
		isLoading,
		totalPages,
		currentPage,
		setCurrentPage,
		showPagination
	} = useProducts(category, search);

	return (
		<CartProvider>
			<Header onSearch={setSearch} onFilterCategory={setCategory} />
			{isLoading ? (
				<Spinner />
			) : (
				<div className="body-content">
					<ProductList products={products} />
					{showPagination ? (
						<Paginator
							currentPage={currentPage}
							totalPages={totalPages}
							onPageChange={setCurrentPage}
						/>
					) : (
						<></>
					)}
				</div>
			)}
			<Footer />
		</CartProvider>
	);
};

export default HomePage;
