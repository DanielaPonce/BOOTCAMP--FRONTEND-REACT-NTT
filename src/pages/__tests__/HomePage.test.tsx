import { jest } from '@jest/globals';
import { fireEvent, render, screen } from '@testing-library/react';
import Footer from '../../components/Footer/Footer';
import Header from '../../components/Header/Header';
import Paginator from '../../components/ProductList/Paginator';
import ProductList from '../../components/ProductList/ProductList';
import Spinner from '../../components/Spinner/Spinner';
import { useProducts } from '../../hooks/useProducts';
import HomePage from '../HomePage';

jest.mock('../../hooks/useProducts');

jest.mock('../../components/Header/Header', () => ({
	__esModule: true,
	default: jest.fn(
		(props: {
			onSearch: (value: string) => void;
			onFilterCategory: (value: string) => void;
		}) => {
			const { onSearch, onFilterCategory } = props;
			return (
				<div>
					<input
						data-testid="search-input"
						onChange={(e) => onSearch(e.target.value)}
					/>
					<select
						data-testid="category-select"
						onChange={(e) => onFilterCategory(e.target.value)}
					>
						<option value="all">All</option>
						<option value="category1">Category 1</option>
					</select>
				</div>
			);
		}
	)
}));

jest.mock('../../components/Footer/Footer', () => ({
	__esModule: true,
	default: jest.fn(() => <footer>Footer</footer>)
}));

jest.mock('../../components/Spinner/Spinner', () => ({
	__esModule: true,
	default: jest.fn(() => <div>Loading...</div>)
}));

jest.mock('../../components/ProductList/ProductList', () => ({
	__esModule: true,
	default: jest.fn((props: { products: any[] }) => {
		const { products } = props;
		return (
			<div>
				{products.map((product) => (
					<div key={product.id}>{product.title}</div>
				))}
			</div>
		);
	})
}));

jest.mock('../../components/ProductList/Paginator', () => ({
	__esModule: true,
	default: jest.fn(
		(props: {
			currentPage: number;
			totalPages: number;
			onPageChange: (page: number) => void;
		}) => {
			const { currentPage, totalPages, onPageChange } = props;
			return (
				<div>
					<button
						data-testid="prev-page"
						onClick={() => onPageChange(currentPage - 1)}
						disabled={currentPage === 1}
					>
						Previous
					</button>
					<span>{`Page ${currentPage} of ${totalPages}`}</span>
					<button
						data-testid="next-page"
						onClick={() => onPageChange(currentPage + 1)}
						disabled={currentPage === totalPages}
					>
						Next
					</button>
				</div>
			);
		}
	)
}));

describe('HomePage Component', () => {
	const mockProducts = [
		{
			id: 1,
			title: 'Product 1',
			description: 'Description for product 1',
			category: 'category1',
			price: 100,
			thumbnail: 'http://example.com/product1.jpg'
		}
	];

	beforeEach(() => {
		jest.clearAllMocks();
	});

	test('renders Header, Footer, and Spinner when loading', () => {
		(useProducts as jest.Mock).mockReturnValue({
			products: [],
			isLoading: true,
			totalPages: 1,
			currentPage: 1,
			setCurrentPage: jest.fn(),
			showPagination: false
		});

		render(<HomePage />);

		expect(Header).toHaveBeenCalled();
		expect(Spinner).toHaveBeenCalled();
		expect(Footer).toHaveBeenCalled();
		expect(ProductList).not.toHaveBeenCalled();
		expect(Paginator).not.toHaveBeenCalled();
	});

	test('renders ProductList and Paginator when not loading', () => {
		const mockSetCurrentPage = jest.fn();

		(useProducts as jest.Mock).mockReturnValue({
			products: mockProducts,
			isLoading: false,
			totalPages: 2,
			currentPage: 1,
			setCurrentPage: mockSetCurrentPage,
			showPagination: true
		});

		render(<HomePage />);

		expect(ProductList).toHaveBeenCalledWith(
			{ products: mockProducts },
			expect.anything()
		);
		expect(Paginator).toHaveBeenCalledWith(
			{
				currentPage: 1,
				totalPages: 2,
				onPageChange: mockSetCurrentPage
			},
			expect.anything()
		);
	});

	test('updates search and category filters', () => {
		const mockSetCurrentPage = jest.fn();

		(useProducts as jest.Mock).mockReturnValue({
			products: mockProducts,
			isLoading: false,
			totalPages: 1,
			currentPage: 1,
			setCurrentPage: mockSetCurrentPage,
			showPagination: false
		});

		render(<HomePage />);

		const searchInput = screen.getByTestId('search-input');
		const categorySelect = screen.getByTestId('category-select');

		fireEvent.change(searchInput, { target: { value: 'Product' } });
		fireEvent.change(categorySelect, { target: { value: 'category1' } });

		expect(Header).toHaveBeenCalled();
	});

	test('paginates correctly', () => {
		const mockSetCurrentPage = jest.fn();

		(useProducts as jest.Mock).mockReturnValue({
			products: mockProducts,
			isLoading: false,
			totalPages: 2,
			currentPage: 1,
			setCurrentPage: mockSetCurrentPage,
			showPagination: true
		});

		render(<HomePage />);

		const nextPageButton = screen.getByTestId('next-page');

		fireEvent.click(nextPageButton);

		expect(mockSetCurrentPage).toHaveBeenCalledWith(2);
	});

	test('matches snapshot', () => {
		(useProducts as jest.Mock).mockReturnValue({
			products: mockProducts,
			isLoading: false,
			totalPages: 1,
			currentPage: 1,
			setCurrentPage: jest.fn(),
			showPagination: false
		});

		const { asFragment } = render(<HomePage />);
		expect(asFragment()).toMatchSnapshot();
	});
});
