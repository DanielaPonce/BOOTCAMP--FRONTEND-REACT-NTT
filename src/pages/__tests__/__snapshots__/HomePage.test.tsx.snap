// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <input
      data-testid="search-input"
    />
    <select
      data-testid="category-select"
    >
      <option
        value="all"
      >
        All
      </option>
      <option
        value="category1"
      >
        Category 1
      </option>
    </select>
  </div>
  <div
    class="body-content"
  >
    <div>
      <div>
        Product 1
      </div>
    </div>
  </div>
  <footer>
    Footer
  </footer>
</DocumentFragment>
`;
